src/app/clusters/actions-server/actions-server.cpp

```diff
@@ -0,0 +1,355 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "actions-server.h"
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app/AttributeAccessInterfaceRegistry.h>
+#include <app/CommandHandlerInterfaceRegistry.h>
+#include <app/EventLogging.h>
+#include <app/util/attribute-storage.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::Actions;
+using namespace chip::app::Clusters::Actions::Attributes;
+using namespace chip::Protocols::InteractionModel;
+
+Instance Instance::instance;
+Instance * Instance::GetInstance()
+{
+    return &instance;
+}
+
+void Instance::OnStateChanged(EndpointId endpoint, uint16_t actionId, uint32_t invokeId, ActionStateEnum actionState)
+{
+    ChipLogProgress(Zcl, "ActionsServer: OnStateChanged");
+
+    // Record StateChanged event
+    EventNumber eventNumber;
+    Events::StateChanged::Type event{ actionId, invokeId, actionState };
+
+    if (CHIP_NO_ERROR != LogEvent(event, endpoint, eventNumber))
+    {
+        ChipLogError(Zcl, "ActionsServer: Failed to record OnStateChanged event");
+    }
+}
+
+void Instance::OnActionFailed(EndpointId endpoint, uint16_t actionId, uint32_t invokeId, ActionStateEnum actionState,
+                              ActionErrorEnum actionError)
+{
+    ChipLogProgress(Zcl, "ActionsServer: OnActionFailed");
+
+    // Record ActionFailed event
+    EventNumber eventNumber;
+    Events::ActionFailed::Type event{ actionId, invokeId, actionState, actionError };
+
+    if (CHIP_NO_ERROR != LogEvent(event, endpoint, eventNumber))
+    {
+        ChipLogError(Zcl, "ActionsServer: Failed to record OnActionFailed event");
+    }
+}
+
+CHIP_ERROR Instance::ReadActionListAttribute(const AttributeValueEncoder::ListEncodeHelper & encoder)
+{
+    if (GetInstance()->mDelegate == nullptr)
+    {
+        ChipLogError(Zcl, "Actions delegate is null!!!");
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+    for (uint16_t i = 0; i <= kMaxActionList; i++)
```

Comment:
Why this kMaxActionList cap?  Just keep going until the provider returns CHIP_ERROR_PROVIDER_LIST_EXHAUSTED?

That said, we could keep this cap to ensure we don't produce invalid input even with a buggy provider, but then I would suggest we should call this cap kMaxActionListLength.